### PR TITLE
[TRTLLM-7028][feat] Enable guided decoding with speculative decoding (part 2: one-model engine)

### PR DESCRIPTION
**Pull request by syuoni**
# [TRTLLM-7028][feat] Enable guided decoding with speculative decoding (part 2: one-model engine)

## Background

When  multiple (draft and target) generation steps are launched by a single CUDA Graph, the CPU-GPU synchronization is difficult:
* For every step `i`, there are two event waits
    * Before grammar advance, the host waits the CPU tokens from step `i-1`
    * Before sampling, the main stream waits the GPU bitmask from step `i`
* If multi-step forwards are launched by a single CUDA Graph, it is not possible to let the CUDA Graph wait events recorded from the host anymore.

CUDA callback [cudaLaunchHostFunc](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__EXECUTION.html#group__CUDART__EXECUTION_1g05841eaa5f90f27124241baafb3e856f) can launch a host function to a CUDA stream, so the host function can be captured and replayed by CUDA graph. Hence, we can launch the grammar compute as a CUDA callback. The CUDA graph should capture and replay multi-step model forwards and grammar compute all together.

image

See https://github.com/NVIDIA/TensorRT-LLM/pull/6414 for more details on the CUDA callback integration to Python.

## Description

This PR supports guided decoding with one-model speculative decoding:
* MTP vanilla
* MTP eagle
* Eagle3 (one-model implementation)

This PR introduces a variant of `GuidedDecoder` -- `CapturableGuidedDecoder`, which is embedded in the modeling code and capturable by CUDA graph.

This PR enables GIL releases for some bindings which are likely to call CUDA APIs. This is to avoid potential deadlock with the CUDA callback. Without this context, releasing GIL should be a recommended practice -- for example, PyTorch almost always releases GIL.

## TODO (Aug. 28)

* ~Fix GIL issues of bindings~
* ~Fix nanobind~
* Optimize/fuse bitmask op (https://github.com/NVIDIA/TensorRT-LLM/pull/7481)
* ~Free hostfunc handles automatically~

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

### CODE REVIEW

**Review by Linda-Stadter — COMMENTED** [review]
I looked at the GIL releases and found two cases where the GIL release is not safe. Let me know what you think

File: tensorrt_llm/_torch/pyexecutor/guided_decoder.py [hunk 1]
```diff
@@ -252,18 +370,161 @@ def rollback_draft_tokens(self,
...
-                        llm_req.guided_decoding_params)
+                    req.seq_slot] = self.grammar_matcher_factory.create(
+                        req.guided_decoding_params)
+
+    @nvtx_range("GuidedDecoder.init_disagg_gen_requests")
+    def init_disagg_gen_requests(self) -> None:
+        self._init_disagg_gen_requests(self.requests)
+
+
+class GuidedWorker(GuidedDecoder):
```
Comment (QiJune) [inline tensorrt_llm/_torch/pyexecutor/guided_decoder.py:395]: `GuidedWorker` is good name, but it doesn't immediately tell users how it's special. Its key feature is that it's compatible with CUDA graphs, and "Worker" is a bit generic. How about renaming it to `CapturableGuidedDecoder`?

File: tests/integration/defs/accuracy/test_disaggregated_serving.py [hunk 2]
```diff
@@ -723,6 +726,59 @@ def test_auto_dtype(self, overlap_scheduler, mtp_nextn):
             task = GSM8K(self.MODEL_NAME)
             task.evaluate(llm)
 
+    @parametrize_with_ids("mtp_nextn",
+                          [0, pytest.param(2, marks=skip_pre_hopper)])
+    @pytest.mark.parametrize("backend", ["xgrammar", "llguidance"])
+    def test_guided_decoding(self, backend: str, mtp_nextn: int, mocker):
+        mocker.patch.dict(os.environ, {"TRTLLM_XGUIDANCE_LENIENT": "1"})
+        ctx_server_config = {
+            "disable_overlap_scheduler": True,
```
Comment (QiJune) [inline tests/integration/defs/accuracy/test_disaggregated_serving.py:735]: Should we enable overlap scheduler for MTP?

File: tensorrt_llm/_torch/pyexecutor/guided_decoder.py [hunk 3]
```diff
@@ -252,18 +370,161 @@ def rollback_draft_tokens(self,
...
+                assert not req.is_draft
+                req.cast_to_draft()
+            else:
+                assert req.is_draft
+
+    def execute_draft_batch(self,
+                            logits: torch.Tensor,
+                            d2t: Optional[torch.Tensor] = None,
+                            is_first_step: bool = False,
+                            is_last_step: bool = False) -> None:
```
Comment (QiJune) [inline tensorrt_llm/_torch/pyexecutor/guided_decoder.py:517]: Can we pass `step` here, instead of `is_first_step`/`is_last_step`?

```python
def execute_draft_batch(self, xxx, step):
    is_first_step = (step == 0)
    is_last_step = step == self.max_num_draft_tokens
    xxx
```

File: tensorrt_llm/_torch/pyexecutor/py_executor_creator.py [hunk 4]
```diff
@@ -393,19 +393,29 @@ def drafting_loop_wrapper(model):
...
                     executor_config.guided_decoding_config,
-                    executor_config.max_batch_size,
-                    model_engine.model.vocab_size_padded,
-                    max_num_draft_tokens=max_num_draft_tokens)
+                    "max_num_sequences": executor_config.max_batch_size,
+                    "vocab_size_padded": model_engine.model.vocab_size_padded
+                }
+                if spec_config is not None:
+                    kwargs["max_num_draft_tokens"] = spec_config.max_draft_len
+                if spec_config is None or spec_config.spec_dec_mode.support_guided_decoder(
```
[the remark this answers is not in the document]
Comment (syuoni) [reply tensorrt_llm/_torch/pyexecutor/py_executor_creator.py:406]: Hi @QiJune , this polished version looks a bit unnatural to me. For example, it uses a statement with side effect in the `if` condition.

For the `use_worker` and `use_decoder`, they are likely to be changed soon. @mikeiovine  introduces capturable drafting loop for the two-model spec dec, so it should require `CapturableGuidedDecoder`.

I added two comments explaining the current use cases for `GuidedDecoder` and `CapturableGuidedDecoder`. Hope this can make it clear, thanks!

File: tensorrt_llm/_torch/pyexecutor/guided_decoder.py (hunk 1 above)
Comment (syuoni) [reply tensorrt_llm/_torch/pyexecutor/guided_decoder.py]: Good suggestion! Renamed `GuidedWorker` to `CapturableGuidedDecoder`.

File: tensorrt_llm/_torch/pyexecutor/guided_decoder.py [hunk 5]
```diff
@@ -252,18 +370,161 @@ def rollback_draft_tokens(self,
...
+        self.queue.put((self.requests, new_tokens is not None))
+        # self.token_event.record() should be called inside CUDA graph capturing;
+        # currently, it is in PyTorchModelEngine._preprocess_inputs.
+
+    @hostfunc
+    def fetch_batch(self) -> None:
+        # CUDA graph warmup calls model forward for multiple times for one prepared inputs
+        if self.queue.empty():
+            return
+        self.requests_hostfunc, has_new_tokens = self.queue.get()
```
[the remark this answers is not in the document]
Comment (syuoni) [reply tensorrt_llm/_torch/pyexecutor/guided_decoder.py:434]: This `xxx_hostfunc` naming is following :
* Tensors with naming `xxx_cpu` or `xxx_host` should be accessed from host
* Tensors with naming `xxx_cuda` or `xxx_device` should be accessed from device

`xxx_hostfunc` might be the third category that should be accessed from hostfunc (CUDA callback). For now, I kept `requests_hostfunc` naming but added comments to explain it.  What do you think?

File: tests/integration/defs/accuracy/test_disaggregated_serving.py (hunk 2 above)
Comment (syuoni) [reply tests/integration/defs/accuracy/test_disaggregated_serving.py:735]: In fact, disagg context instances should not enable overlap scheduler.

File: tensorrt_llm/_torch/models/modeling_speculative.py [hunk 6]
```diff
@@ -466,3 +465,8 @@ def load_draft_weights(self,
         self.draft_model.load_weights(weights=weights,
                                       weight_mapper=weight_mapper)
         self.draft_model.load_weights_from_target_model(self)
+
+    def set_guided_worker(self, guided_worker: GuidedWorker) -> bool:
```
[the remark this answers is not in the document]
Comment (syuoni) [reply tensorrt_llm/_torch/models/modeling_speculative.py]: Yes, as explained by Michael, this is on purpose.

The return value of this method depends on whether `spec_worker` supports `set_guided_decoder`.

For now, because this PR implements `set_guided_decoder` for all possible kinds of `spec_worker`, so yes, this method always return `True`. But in future, if some kind of `spec_worker` cannot implement `set_guided_decoder`, then this method will return `False`.

File: tensorrt_llm/_torch/pyexecutor/guided_decoder.py [hunk 7]
```diff
@@ -65,115 +185,96 @@ def __init__(self,
...
         """Build the bitmask for requests with guided decoding enabled.
 
         Specifically, this method:
         - build and advance the grammar matcher for context and generation requests, respectively;
         - call the grammar matcher to fill the bitmask on CPU;
         - asynchronously copy the bitmask to GPU.
         """
-        for llm_req in scheduled_requests.all_requests():
-            slot: int = llm_req.py_target_seq_slot if llm_req.py_is_draft else llm_req.py_seq_slot
+        # Fix it.
```
[the remark this answers is not in the document]
Comment (syuoni) [reply tensorrt_llm/_torch/pyexecutor/guided_decoder.py]: Added, thanks for the reminder. In fact, the fix is on the way: https://github.com/NVIDIA/TensorRT-LLM/pull/7481

File: tensorrt_llm/_torch/pyexecutor/guided_decoder.py (hunk 3 above)
Comment (syuoni) [reply tensorrt_llm/_torch/pyexecutor/guided_decoder.py]: Great suggestion! Revised accordingly, this simplifies the API a lot, thanks!

File: docs/source/torch/features/feature_combination_matrix.md [hunk 8]
```diff
@@ -15,4 +15,4 @@
...
 | Slide Window Attention     | Yes               | Yes        | Yes                        | Yes                        | Yes              | No       | Untested                  | Untested                  | Yes           | Yes              | WIP            | ---                    |                       |                 |
 | Logits Post Processor      | Yes                | Yes        | Yes                        | No                    | Yes             | No       | No                        | No                        | Yes           | Yes              | Yes            | Yes                    | ---                   |                 |
-| Guided Decoding            | Yes               | Yes        | Yes                        | Yes                   | Yes             | No       | No                        | Yes                       | Yes           | Yes              | Yes            | Yes                    | Yes                   | ---             |
```
[the remark this answers is not in the document]
Comment (syuoni) [reply docs/source/torch/features/feature_combination_matrix.md:18]: Thanks for the suggestion! This deserves a dedicated PR if we decide to do so.

cc @QiJune  what do you think?

File: cpp/tensorrt_llm/nanobind/runtime/hostfunc.cpp [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+}
+
+void freeHostFuncUserData(uintptr_t userDataPtr)
+{
+    // Acquire the GIL to safely release the Python objects.
+    nb::gil_scoped_acquire gil;
+
+    // Create a unique_ptr to take over the ownership of the user data;
+    // the user data is released when the unique_ptr is destroyed.
+    auto hostFuncUserData = std::unique_ptr<HostFuncUserData>(reinterpret_cast<HostFuncUserData*>(userDataPtr));
```
[the remark this answers is not in the document]
Comment (syuoni) [reply cpp/tensorrt_llm/nanobind/runtime/hostfunc.cpp:104]: I was warned to avoid using C-style code.

I agree that just `delete` is clearer here. If the current code is functionally correct (and equivalent), can we keep it?

File: docs/source/torch/features/feature_combination_matrix.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (QiJune) [reply docs/source/torch/features/feature_combination_matrix.md:18]: Yes, it can be updated with another PR.

**Review by QiJune — APPROVED** [review]
LGTM

File: cpp/tensorrt_llm/nanobind/batch_manager/kvCacheManager.cpp [hunk 10]
```diff
@@ -335,23 +335,28 @@ void tb::kv_cache_manager::KVCacheManagerBindings::initBindings(nb::module_& m)
         .def_static("calculate_max_num_blocks", &tbk::BaseKVCacheManager::calculateMaxNumBlocks, nb::arg("config"),
             nb::arg("is_cross_attention"), nb::arg("dtype"), nb::arg("model_config"), nb::arg("world_config"),
             nb::arg("window_size_to_layers"), nb::arg("allotted_primary_mem_bytes"),
-            nb::arg("allotted_secondary_mem_bytes"), nb::arg("extra_cost_memory"), nb::arg("kv_factor"))
-        .def("allocate_pools", &BaseKVCacheManager::allocatePools)
-        .def("release_pools", &BaseKVCacheManager::releasePools)
-        .def("start_scheduling", &BaseKVCacheManager::startScheduling)
+            nb::arg("allotted_secondary_mem_bytes"), nb::arg("extra_cost_memory"), nb::arg("kv_factor"),
+            nb::call_guard<nb::gil_scoped_release>())
+        .def("allocate_pools", &BaseKVCacheManager::allocatePools, nb::call_guard<nb::gil_scoped_release>())
```
[the remark this answers is not in the document]
Comment (syuoni) [reply cpp/tensorrt_llm/nanobind/batch_manager/kvCacheManager.cpp:340]: Thanks for pointing this out. As offline discussed, nanobind/pybind [acquires GIL again before entering the Python code](https://github.com/wjakob/nanobind/blob/master/src/trampoline.cpp#L69), so it's safe to add `py::call_guard<py::gil_scoped_release>()` for these pure virtual functions.

File: cpp/tensorrt_llm/nanobind/batch_manager/kvCacheManager.cpp [hunk 11]
```diff
@@ -474,35 +497,39 @@ void tb::kv_cache_manager::KVCacheManagerBindings::initBindings(nb::module_& m)
...
             nb::arg("event_manager") = nullptr, nb::arg("enable_partial_reuse") = true,
-            nb::arg("copy_on_partial_reuse") = true, nb::arg("kv_connector_manager") = nullptr);
+            nb::arg("copy_on_partial_reuse") = true, nb::arg("kv_connector_manager") = nullptr,
+            nb::call_guard<nb::gil_scoped_release>());
 }
 
 void tb::BasePeftCacheManagerBindings::initBindings(nb::module_& m)
 {
     nb::class_<tb::BasePeftCacheManager, PyBasePeftCacheManager>(m, "BasePeftCacheManager")
         .def("add_request_peft", &tb::BasePeftCacheManager::addRequestPeft, nb::arg("request"),
```
[the remark this answers is not in the document]
Comment (syuoni) [reply cpp/tensorrt_llm/nanobind/batch_manager/kvCacheManager.cpp:507]: Replied in another thread, thanks.

**Review by Linda-Stadter — APPROVED** [review]

**Review by mikeiovine — APPROVED** [review]
